### PR TITLE
qt 5.12.0 -> 5.12.3 WIP

### PR DESCRIPTION
**Pull request by vandenoever**
###### Motivation for this change

Qt 5.12 has received 3 bugfix releases since 5.12.0. However updating it has proven non-trivial. This PR  is a WIP.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by vandenoever** [comment]
See also https://github.com/NixOS/nixpkgs/issues/57042

**Comment by ttuegel** [comment]
Does the configure script produce a log file, so we could see what compilation tests it's doing?

**Comment by vandenoever** [comment]
Yes there was.

[config.log](https://github.com/NixOS/nixpkgs/files/3124296/config.log)

**Comment by vandenoever** [comment]
qtbase compiles in the latest version of this patch.

**Comment by vandenoever** [comment]
The current patch compiles, but does not work with X11: a KDE system with it does not start the graphical login `sddm`.

This is the configuration report of Qt 5.12.0 next to that of Qt 5.12.3:
```
Qt Gui:                                                         Qt Gui:
  Accessibility .......................... yes                    Accessibility .......................... yes
  FreeType ............................... yes                    FreeType ............................... yes
    Using system FreeType ................ yes                      Using system FreeType ................ yes
  HarfBuzz ............................... yes                    HarfBuzz ............................... yes
    Using system HarfBuzz ................ yes                      Using system HarfBuzz ................ yes
  Fontconfig ............................. yes                    Fontconfig ............................. yes
  Image formats:                                                  Image formats:
    GIF .................................. yes                      GIF .................................. yes
    ICO .................................. yes                      ICO .................................. yes
    JPEG ................................. yes                      JPEG ................................. yes
      Using system libjpeg ............... yes                        Using system libjpeg ............... yes
    PNG .................................. yes                      PNG .................................. yes
      Using system libpng ................ yes                        Using system libpng ................ yes
  EGL .................................... yes                    EGL .................................... yes
  OpenVG ................................. no                     OpenVG ................................. no
  OpenGL:                                                         OpenGL:
    Desktop OpenGL ....................... yes                      Desktop OpenGL ....................... yes
    OpenGL ES 2.0 ........................ no                       OpenGL ES 2.0 ........................ no
    OpenGL ES 3.0 ........................ no                       OpenGL ES 3.0 ........................ no
    OpenGL ES 3.1 ........................ no                       OpenGL ES 3.1 ........................ no
    OpenGL ES 3.2 ........................ no                       OpenGL ES 3.2 ........................ no
  Vulkan ................................. no                     Vulkan ................................. no
  Session Management ..................... yes                    Session Management ..................... yes
Features used by QPA backends:                                  Features used by QPA backends:
  evdev .................................. yes                    evdev .................................. yes
  libinput ............................... yes                    libinput ............................... yes
  INTEGRITY HID .......................... no                     INTEGRITY HID .......................... no
  mtdev .................................. no                     mtdev .................................. no
  tslib .................................. no                     tslib .................................. no
  xkbcommon-evdev ........................ yes                |   xkbcommon .............................. yes
                                                              >   X11 specific:
                                                              >     XLib ................................. no
                                                              >     XCB Xlib ............................. no
                                                              >     EGL on X11 ........................... no
QPA backends:                                                   QPA backends:
  DirectFB ............................... no                     DirectFB ............................... no
  EGLFS .................................. no                     EGLFS .................................. no
  LinuxFB ................................ no                     LinuxFB ................................ no
  VNC .................................... yes                    VNC .................................... yes
  Mir client ............................. no                     Mir client ............................. no
  X11:                                                        |   XCB:
    Using system-provided XCB libraries .. yes                      Using system-provided XCB libraries .. yes
    EGL on X11 ........................... yes                <
    XCB XKB .............................. yes                      XCB XKB .............................. yes
    XLib ................................. yes                <
    XCB render ........................... yes                <
    XCB GLX .............................. yes                <
    XCB XInput ........................... yes                      XCB XInput ........................... yes
    XCB Xlib ............................. yes                |     Native painting (experimental) ....... no
    Using system-provided xkbcommon ...... yes                |     GL integrations:
    Native painting (experimental) ....... yes                |       GLX Plugin ......................... no
                                                              >       EGL-X11 Plugin ..................... no
```

The GLX Plugin and EGL-X11 Plugin are missing, as is the XLib detection.

**Comment by vandenoever** [comment]
The current version of the patch detects X properly. All the features detected in Qt 5.12.0 are also detected in Qt 5.12.3.

Testing will be done once the patched version and KDE software has compiled.

**Comment by vandenoever** [comment]
I'm running my desktop now with Qt 5.12.3 (this patch) and the latest KDE software on NixOS 19.03.

Everything works except the global KWin shortcuts (Alt-Tab, Ctrl-F1 etc). Other global shortcut like Alt-Space / Alt-F2, Meta-P still work. I'm not sure this is related to this patch.

**Comment by vandenoever** [comment]
After rebooting the machine the global kwin shortcuts work again. In fact, the bug that I hoped would be fixed by upgrading from Qt 5.12.0 to Qt 5.12.3 is gone. https://bugs.kde.org/show_bug.cgi?id=406829

**Comment by dtzWill** [comment]
On Mon, 06 May 2019 16:13:59 -0700, Jos van den Oever <notifications@github.com> wrote:
> After rebooting the machine the global kwin shortcuts work again. In fact, the bug that I hoped would be fixed by upgrading from Qt 5.12.0 to Qt 5.12.3 is gone. https://bugs.kde.org/show_bug.cgi?id=406829

I don't run KDE, but use lots of qt-based/kde software -- and so far
this is working great here as well! <3

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -12717,7 +12717,7 @@ in
       inherit stdenv fetchurl fetchFromGitHub makeSetupHook;
       bison = bison2; # error: too few arguments to function 'int yylex(...
       inherit cups;
-      harfbuzz = harfbuzzFull;
+      inherit harfbuzz;
```
Comment (eadwu) [inline pkgs/top-level/all-packages.nix]: This should be under `qt512` in order for `qtbase` to pass `configurePhase`.

**Review by eadwu — CHANGES_REQUESTED** [review]

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (vandenoever) [reply pkgs/top-level/all-packages.nix]: Fixed!

**Comment by eadwu** [comment]
Builds fine in sandbox on `x86_64`. Everything I use looks more or less the same with the upgrade while what was broken with `5.12.0` is still broken for me.

**Comment by vandenoever** [comment]
In the meantime qt 5.12.4 has been released.

**Comment by dtzWill** [comment]
5.12.4 seems close to 5.12.3, but I'm having trouble "sanely" forward-porting the `qtbase`-related patches-- mostly I'm at the point of essentially rewriting them from scratch (so that I understand the pieces, how they've changed, and how we may want them to work in NixOS-land).  Unfortunately I am not up to this task for the coming weeks (months) most likely, but maybe others find better success with the changes (this is where my hope lies :)).  I don't think it's too hairy, but may require better quickfix-fu or some degree of deep-diving.  It's possible 5.12.4 moves things more towards our preferred situation... that I can't tell is why I'm writing this comment and calling it quits for a bit O:).

**Comment by dtzWill** [comment]
I'm not sure we would be able to move to it piecemeal, but notably qt 5.12.4 supports openssl 1.1 which might motivate sorting these things out eventually :).

**Review by ttuegel — COMMENTED** [review]
This looks good to me! We can worry about Qt 5.12.4 later. I'm not going to "approve" because this should go into `staging`; @vandenoever would you mind opening a new pull request? Sorry :slightly_frowning_face:

**Comment by FRidh** [comment]
Closing in favor of https://github.com/NixOS/nixpkgs/pull/64598

**Pull request closed without merging**